cosmology/GravityManager.h

```diff
@@ -0,0 +1,367 @@
+
+    /**
+     * @brief Get the current simulation time.
+     * 
+     * @return Current simulation time.
+     */
+    double getTime() const { return time_m; }
+
+    /**
+     * @brief Set the current simulation time.
+     * 
+     * @param time_ Current simulation time.
+     */
+    void setTime(double time_) { time_m = time_; }
+
+    /**
+     * @brief Set the initial conditions folder.
+     * 
+     * @param ic_folder Initial conditions folder.
+     */
+    void setIC(std::string ic_folder) {folder = ic_folder;}
+
+    // Calculation methods
+
+    /**
+     * @brief Calculate the time given a scaling factor.
+     * 
+     * @param a Scaling factor.
+     * @return Calculated time.
+     */
+    double calculateTime(double a) {
+        return this->t_L * asinh(sqrt(pow(a, 3)* this->O_L / this->O_m)); // inverse function of calculateScaling
+    }
+    
+    /**
+     * @brief Calculate the scaling factor given a time.
+     * 
+     * @param t Time.
+     * @return Calculated scaling factor.
+     */
+    double calculateScaling(double t){
+        return pow(this->O_m/this->O_L, 1./3.)*pow(sinh(t/this->t_L), 2./3.); // https://arxiv.org/pdf/0803.0982.pdf (p. 6)
+    }
+    
+    /**
+     * @brief Calculate the Hubble parameter given a scaling factor.
+     * 
+     * @param a Scaling factor.
+     * @return Calculated Hubble parameter.
+     */
+    double calculateHubble(double a){
+        return this->Hubble0 * sqrt( this->O_m / pow(a, 3) + this->O_L );
+    }
+
+    /**
+     * @brief Initialize the simulation time and related parameters.
+     */
+    void InitialiseTime(){
+        Inform mes("Inititalise: ");
+        this->O_m = 0.3;
+        this->O_L = 0.7;
+        this->t_L = 2/(3*this->Hubble0*sqrt(this->O_L));
+        this->a_m = 1/(1+this->z_m);
+        this->Dloga = 1. /(this->nt_m) * log((1+this->z_m)/(1+this->z_f));
+        this->time_m = this->calculateTime(this->a_m);
+        this->Hubble_m = this->calculateHubble(this->a_m); // Hubble parameter at starting time
+        this->dt_m = this->Dloga / this->Hubble_m;
+        this->rho_crit0 = 3 * this->Hubble0 * this->Hubble0 / (8*M_PI * this->G); // critical density today
+
+        // Print initial parameters
+        mes << "time: " << this->time_m << ", timestep: " << this->dt_m << endl;
+        mes << "Dloga: " << this->Dloga << endl;
+        mes << "z: " << this->z_m << ", scaling factor: " << this->a_m << endl;
+        mes << "H0: " << this->Hubble0 << ", H_initial: " << this->Hubble_m << endl;
+        mes << "critical density (today): " << this->rho_crit0 << endl;
+    }
+
+    /**
+     * @brief Virtual method to dump data (default does nothing).
+     */
+    virtual void dump() { /* default does nothing */ };
+
+    // Step methods
+
+    /**
+     * @brief Pre-step method called before each simulation step.
+     */
+    void pre_step() override {
+        Inform mes("Pre-step");
+        mes << "Done" << endl;
+    }
```

Comment:
maybe a comment indicating why it does nothing would be informative

Reply:
I'm not sure, but is it just because it is not necessary and because it's an inherited function? Why just not skip it in that case?

Reply:
Actually, just noticed this is the same in Alpine... I am not sure what the purpose of this is (@srikrrish ?). But I guess we leave it like this since it is the same there.

Reply:
Yeah the PIC manager is very generic and not all the applications need all steps. In case of Alpine the pre_step (that is something before a time step is taken (Note: it is different from pre_run which is initialization)) was not needed. May be it is the same case here too.